Add coding sequence generator

arendsee opened the pull request:
* I'm not sure if the documentation syntax is correct
 * no handling in coding sequence generator for ambiguous bases
 * should we implement a `max_madness` function that generates that
 captures the variety we see in the wild?

arendsee committed:
Add coding sequence generator

 * I'm not sure if the documentation syntax is correct
 * no handling in coding sequence generator for ambiguous bases
 * should we implement a `max_madness` function that generates that
 captures the variety we see in the wild?

luizirber commented:
@arendsee draft PRs don't get build by travis, so I marked this ready to review. But take your time =]

arendsee commented:
@luizirber I'm not quite sure about the workflow, is there something that I need to do?

luizirber committed:
Merge branch 'master' into devZ

luizirber commented:
I resolved the merge conflicts.

As of #3 the dev tasks are defined in tox, so running
```
$ tox
```
will do all lint checks (and format code using black), and also runs tests in py{35,36,37} (and skip missing interpreters). This is all heavily based on how tox runs dev tasks.

luizirber committed:
fix lint errors

luizirber committed:
add codecov env vars

luizirber commented:
And now it's failing because there is new code with is not tested =]

Zac-HD commented on hypothesis_bio/strategies.py:
```diff
@@ -149,3 +150,196 @@ def sequence(draw, *, alphabet: Alphabet = Alphabet.DNA, max_size: int = 1000):
+        a single stop codon
+    ----------
+    """
+    return draw(st.sampled_from(stop_codons))
```
You don't need to use `@st.composite` for this at all!  Ditch the overhead by simply `return st.sampled_from(stop_codons)`.

(You may also want to check that everything in the `stop_codons` argument is in fact a valid codon)

Zac-HD commented on hypothesis_bio/strategies.py:
```diff
@@ -149,3 +150,196 @@ def sequence(draw, *, alphabet: Alphabet = Alphabet.DNA, max_size: int = 1000):
+
+
+@st.composite
+def codon(draw, *, stop_codons=CANONICAL_STOP_CODONS):
```
Again, you don't need `@st.composite` for this.  Plus the `stop_codons` argument is unused?

Zac-HD commented on hypothesis_bio/strategies.py:
```diff
@@ -149,3 +150,196 @@ def sequence(draw, *, alphabet: Alphabet = Alphabet.DNA, max_size: int = 1000):
+    ----------
+    """
+    cdn = draw(codon())
+    assume(cdn not in stop_codons)
```
Same as above, with the twist that Hypothesis can do much better via the `.filter` method than `assume`.

So `return codon().filter(lambda c: c not in stop_codons)` would be best 😄

Zac-HD commented on hypothesis_bio/strategies.py:
```diff
@@ -149,3 +150,196 @@ def sequence(draw, *, alphabet: Alphabet = Alphabet.DNA, max_size: int = 1000):
+        If True, allow stop codons inside the codings sequence
+
+    **kwargs
+        Additional arguments passed to :mod:`codons`
```
They're actually passed to `st.lists()`...

And it's possible to define this one entirely without `@st.composite` too, if you think about it carefully:

```python
return st.builds(
    "{}{}{}".format,
    st.just("ATG" if include_start_codon else ""),
    st.lists(codon() if include_stop_codon else non_stop_codon(), **kwargs).map("".join),
    stop_codon() if include_stop_codon else st.just(""),
)
```

Plus some carefully juggling of `**kwargs`` for size if you want it to be correct with respect to the possible start and end codons.

arendsee replied:
Thanks for the review @Zac-HD ! I'm pushing a couple more commits that I believe address everything you mention.

Zac-HD reviewed (commented):
I hope this is helpful - I've included some performance-oriented feedback from a Hypothesis maintainer's perspective, which should also make the code easier to read and document.

arendsee committed:
Remove unnecessary composition decorators

arendsee committed:
Removed unused arguments and fixed documentation

arendsee committed:
Adjust min_ and max_size arguments

Now the min and max size refer to the number of coding amino acids in
the output coding sequence. That is, the STOP codon is not counted. The
START codon is counted, though.

How should the input `min_size=0` be handled when
`include_start_codon=True`? These are contradictory. If there is a START
codon, then the minimum number of codons is 1. Currently, I silently
ignore the 0, treating it as if it were 1 by still including the START
codon.

arendsee committed:
Merge branch 'devZ' of https://github.com/arendsee/hypothesis-bio into devZ

Zac-HD commented:
LGTM, though two notes on docs: you still have the `draw` argument documented in a few places, and strictly speaking these functions return a *strategy for* strings, not a string directly.

arendsee committed:
Remove unused parameters from docs

arendsee committed:
Fix return documentation of returned strategies

arendsee commented:
@Zac-HD Right, I've fixed both issues with the docs.
